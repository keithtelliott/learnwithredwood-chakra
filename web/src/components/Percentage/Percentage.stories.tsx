// See https://storybook.js.org/docs/react/writing-stories/args.

import type { StoryObj, ComponentMeta } from '@storybook/react'

import { Percentage, type PercentageProps } from './Percentage'

export default {
  title: 'Components/Percentage',
  component: Percentage,
} as ComponentMeta<typeof Percentage>

type Story = StoryObj<typeof Percentage>

export const Generated: Story = {
  args: { percentage: 66 } as PercentageProps,
}

export const NoRing: Story = {
  args: { ...Generated.args, showRing: false } as PercentageProps,
}

export const Small: Story = {
  args: {
    ...Generated.args,
    showRing: false,
    size: 'small',
  } as PercentageProps,
}

export const Medium: Story = {
  args: {
    ...Small.args,
    size: 'medium',
  } as PercentageProps,
}

export const Large: Story = {
  args: {
    ...Small.args,
    size: 'large',
  } as PercentageProps,
}

export const XLarge: Story = {
  args: {
    ...Small.args,
    size: 'xlarge',
  } as PercentageProps,
}

export const LowRating: Story = {
  args: {
    ...Small.args,
    showRing: true,
    size: 'medium',
    percentage: 30,
  } as PercentageProps,
}

export const MidRating: Story = {
  args: {
    ...LowRating.args,
    percentage: 60,
  } as PercentageProps,
}

export const HighRating: Story = {
  args: {
    ...LowRating.args,
    percentage: 90,
  } as PercentageProps,
}
