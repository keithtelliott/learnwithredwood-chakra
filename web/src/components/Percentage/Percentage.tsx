import { AbsoluteCenter, Box, HStack, Text } from '@chakra-ui/react'

import { PercentSign } from './PercentSign'

interface Props {
  percentage: number
  showRing?: boolean
  size?: 'small' | 'medium' | 'large' | 'xlarge'
}

const Percentage = ({ percentage, showRing = true, size = 'small' }: Props) => {
  /**
   * KTE, 4/26/2023:  This Chakra implementation works, but I'm sure there's a more-polished way to do it.
   * However, I'm calling it good for now...
   */

  const ringSizes = {
    small: 110,
    medium: 300,
    large: 0,
    xlarge: 0,
  }

  return (
    <Box
      position="relative"
      width={showRing && ringSizes[size] > 0 ? ringSizes[size] + 'px' : '100%'}
      height={showRing && ringSizes[size] > 0 ? ringSizes[size] + 'px' : '100%'}
      color={determineColor(percentage)}
    >
      {showRing && ringSizes[size] > 0 ? (
        <>
          <svg
            data-testid="ring"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/svg"
            width={ringSizes[size]}
            height={ringSizes[size]}
            viewBox={'0 0 230 230'}
          >
            <circle
              cx="115"
              cy="115"
              r="100"
              stroke="#393939"
              strokeWidth={8}
              fill="black"
            />
            <circle
              data-testid="percentageRing"
              cx="115"
              cy="115"
              r="100"
              strokeWidth={8}
              fill="none"
              // className="stroke-current" // KTE, 5/1/2023: This works if Tailwind is installed, but I'll use the following stroke color instead, which jives with Chakra
              stroke={determineColor(percentage)}
              strokeDasharray={`calc(628.32 * ${percentage} / 100) 628.32`}
              transform="rotate(-90) translate(-230)"
            />
          </svg>
          <AbsoluteCenter>
            <NumberAndSymbol
              percentage={percentage}
              size={size}
              color={determineColor(percentage)}
            />
          </AbsoluteCenter>
        </>
      ) : (
        <Box>
          <NumberAndSymbol
            percentage={percentage}
            size={size}
            color={determineColor(percentage)}
          />
        </Box>
      )}
    </Box>
  )
}

const determineColor = (percentage: number) => {
  if (percentage < 50) {
    return 'red'
  } else if (50 <= percentage && percentage < 70) {
    // return 'oriolesOrange'
    return 'orange' // Go-Do, KTE, 5/21/2023:  Work through Storybook issues with custom colors.  My hunch is that custom colors may work in the app, but at the moment they are no working in Storybook.
  } else {
    return 'green'
  }
}

const determineFontSize = (size: Props['size']) => {
  if (size === 'small') {
    return '5xl'
  } else if (size === 'medium') {
    return '9.25rem'
  } else if (size === 'large') {
    return '10rem'
  } else {
    return '16rem'
  }
}

const NumberAndSymbol = ({
  percentage,
  size,
  color,
}: {
  percentage: number
  size: Props['size']
  color: string
}) => (
  <HStack>
    <Text
      fontSize={determineFontSize(size)}
      fontFamily="modern"
    >{`${percentage}`}</Text>
    <Box position="relative" top="-3">
      <PercentSign size={size} color={color} />
    </Box>
  </HStack>
)

export { Percentage, Props as PercentageProps }
