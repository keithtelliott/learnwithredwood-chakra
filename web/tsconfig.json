{
  "compilerOptions": {
    // "maxTupleElements": 10, // added by KTE on 5/6/2023
    // "maxUnionSize": 10, // added by KTE on 5/6/2023
    // "maxRecursionDepth": 10, // added by KTE on 5/6/2023
    "noEmit": true,
    "allowJs": true,
    "esModuleInterop": true,
    "target": "esnext",
    "module": "esnext",
    "moduleResolution": "node",
    "baseUrl": "./",
    // "skipLibCheck": false,
    "skipLibCheck": true, // added by KTE on 5/6/2023
    "rootDirs": [
      "./src",
      "../.redwood/types/mirror/web/src",
      "../api/src",
      "../.redwood/types/mirror/api/src"
    ],
    "paths": {
      "src/*": [
        "./src/*",
        "../.redwood/types/mirror/web/src/*",
        "../api/src/*",
        "../.redwood/types/mirror/api/src/*"
      ],
      "$api/*": [ "../api/*" ],
      "types/*": ["./types/*", "../types/*"],
      "@redwoodjs/testing": ["../node_modules/@redwoodjs/testing/web"]
    },
    "typeRoots": ["../node_modules/@types", "./node_modules/@types"],
    "types": ["jest", "@testing-library/jest-dom"],
    "jsx": "preserve",
  },
  "include": [
    "src",
    "../.redwood/types/includes/all-*",
    "../.redwood/types/includes/web-*",
    "../types",
    "./types"
  ]
}
