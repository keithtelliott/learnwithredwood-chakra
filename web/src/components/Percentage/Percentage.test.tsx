import { color } from '@chakra-ui/react'

import { render, screen } from '@redwoodjs/testing/web'

import { Percentage } from './Percentage'

//   Improve this test with help from the Redwood Testing Doc:
//    https://redwoodjs.com/docs/testing#testing-components

describe('Percentage', () => {
  it('renders successfully', () => {
    expect(() => {
      render(<Percentage percentage={0} />)
    }).not.toThrow()
  })

  it('displays the ring', () => {
    render(<Percentage percentage={0} showRing={true} />)
    expect(screen.getByTestId('ring')).toBeInTheDocument()
  })
  it('hides the ring', () => {
    render(<Percentage percentage={0} showRing={false} />)
    expect(screen.queryByTestId('ring')).not.toBeInTheDocument()
  })
  it('renders the correct ring length for a low percentage', () => {
    render(<Percentage percentage={30} showRing={true} />)
    expect(screen.getByTestId('percentageRing')).toHaveAttribute(
      'stroke-dasharray',
      'calc(628.32 * 30 / 100) 628.32'
    )
    expect(screen.getByTestId('percentageRing')).toMatchSnapshot()
  })

  it('renders the correct ring length for a mid percentage', () => {
    render(<Percentage percentage={50} showRing={true} />)
    expect(screen.getByTestId('percentageRing')).toHaveAttribute(
      'stroke-dasharray',
      'calc(628.32 * 50 / 100) 628.32'
    )
    expect(screen.getByTestId('percentageRing')).toMatchSnapshot()
  })
  it('renders the correct ring length for a high percentage', () => {
    render(<Percentage percentage={89} showRing={true} />)
    expect(screen.getByTestId('percentageRing')).toHaveAttribute(
      'stroke-dasharray',
      'calc(628.32 * 89 / 100) 628.32'
    )
    expect(screen.getByTestId('percentageRing')).toMatchSnapshot()
  })
  it('renders the numeric percentage', () => {
    render(<Percentage percentage={89} showRing={true} />)
    expect(screen.getByText('89')).toBeInTheDocument()
  })

  it('renders a small size', () => {
    render(<Percentage percentage={89} showRing={true} size="small" />)
    expect(screen.getByTestId('ring')).toHaveAttribute('height', '110')
  })
  it('renders a medium size', () => {
    render(<Percentage percentage={89} showRing={true} size="medium" />)
    expect(screen.getByTestId('ring')).toHaveAttribute('height', '300')
  })
  it('renders a large size', () => {
    render(<Percentage percentage={89} showRing={true} size="large" />)
    expect(screen.queryByTestId('ring')).not.toBeInTheDocument()
  })
  it('renders a xLarge size', () => {
    render(<Percentage percentage={89} showRing={true} size="xlarge" />)
    expect(screen.queryByTestId('ring')).not.toBeInTheDocument()
  })
  it('renders a low percentage with the correct color', () => {
    render(<Percentage percentage={10} showRing={true} size="medium" />)
    expect(screen.getByTestId('percentageRing')).toHaveAttribute(
      'stroke',
      'red'
    )
  })
  it('renders a mid percentage with the correct color', () => {
    render(<Percentage percentage={50} showRing={true} size="medium" />)
    expect(screen.getByTestId('percentageRing')).toHaveAttribute(
      'stroke',
      'orange'
    )
  })
  it('renders a high percentage with the correct color', () => {
    render(<Percentage percentage={90} showRing={true} size="medium" />)
    expect(screen.getByTestId('percentageRing')).toHaveAttribute(
      'stroke',
      'green'
    )
  })
})
