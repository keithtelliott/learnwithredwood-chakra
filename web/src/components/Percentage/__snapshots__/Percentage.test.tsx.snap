// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Percentage renders the correct ring length for a high percentage 1`] = `
<circle
  cx="115"
  cy="115"
  data-testid="percentageRing"
  fill="none"
  r="100"
  stroke="green"
  stroke-dasharray="calc(628.32 * 89 / 100) 628.32"
  stroke-width="8"
  transform="rotate(-90) translate(-230)"
/>
`;

exports[`Percentage renders the correct ring length for a low percentage 1`] = `
<circle
  cx="115"
  cy="115"
  data-testid="percentageRing"
  fill="none"
  r="100"
  stroke="red"
  stroke-dasharray="calc(628.32 * 30 / 100) 628.32"
  stroke-width="8"
  transform="rotate(-90) translate(-230)"
/>
`;

exports[`Percentage renders the correct ring length for a mid percentage 1`] = `
<circle
  cx="115"
  cy="115"
  data-testid="percentageRing"
  fill="none"
  r="100"
  stroke="orange"
  stroke-dasharray="calc(628.32 * 50 / 100) 628.32"
  stroke-width="8"
  transform="rotate(-90) translate(-230)"
/>
`;
